### ISSUE TITLE
Feature: install_* to respect path of files

### ISSUE THREAD

**Issue by RobinMcCorkell** [mesonbuild/meson#14]
It would be of great benefit if Meson supported something similar to GNU Automake's `nobase` option for installing files (see http://www.gnu.org/software/automake/manual/html_node/Alternative.html).

**Comment by tristan957** [comment]
Why was this closed? `install_subdir` is just a work around. `no_base` is a feature that should be implemented. I don't want Meson flattening my headers on install. It is currently throwing off my entire library because the headers themselves use the nested paths of my project, but when they install, they are flat and thus the includes don't resolve. This is a huge bug in Meson in my opinion. Following the directory structure should be the default. Please reopen the issue.

**Comment by FFY00** [comment]
My workaround:
```meson
# Install files from sources array
# Loop needed as meson doesn't preserve the path for some reason
# See: https://github.com/mesonbuild/meson/issues/3371
include_subdir = 'mylib'
foreach file : mylib_sources
    file_path = file.split('/')

    # Extract folder
    folder_path = []
    foreach path : file_path
        if path != file_path[-1]
            folder_path += path
        endif
    endforeach

    folder = join_paths(folder_path)
    install_headers(file, subdir: join_paths(include_subdir, folder))
endforeach
```

**Comment by nirbheek** [comment]
> I don't want Meson flattening my headers on install.

That sounds like a bug. I don't believe that's the intention of the call. Could you please open a new issue about that?

**Comment by tristan957** [comment]
@nirbheek I believe #3371 is about this behavior.

**Comment by dzhang-b** [comment]
I think there should be two ways of install headers, similar to what described here.
[nobase_headers](https://www.gnu.org/software/automake/manual/html_node/Headers.html)

Currently we have `install_subdir`, but it is not obvious that it should be used to install header if I am not looking at this github issue.

**Comment by eli-schwartz** [comment]
This ticket was originally about having a nobase-style approach to the install_headers() function. Despite adding install_subdir, this did not actually answer the request, even if it could be used that way (in fact, many, many install_subdir uses have nothing to do with headers and people don't always realize it can be used for this).

It also doesn't work when there are some private headers and some public headers.

Anyway, it is now implemented in #10299 and will be available in the next version of Meson as `install_headers('list', 'of', 'headers', preserve_path: true)`
